---
title: "information_about_RNA-seq_DEG.Rmd"
author: "Jacques van Helden"
date: '`r Sys.Date()`'
output:
  html_document:
    fig_caption: yes
    highlight: zenburn
    theme: cerulean
    toc: yes
    toc_depth: 3
    toc_float: true
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 3
  word_document: default
bibliography: bibliography_rna-seq.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Context of the analysis


## Conditions

You dispose of two or more groups of samples, each corresponding to a given **"condition"**. For the sake of simplicity, we will start with two-group comparisons. Multi-group comparisons can be done, but rely on different methods.

Conditions can represent many different things. 

- genotype (for example in Gierliski 2015, we have two genotypes: WT and Snf2 mutant);
- disease versus healthy (for example in Li 2014, there are 92 samples of skin with psoriasis, and 82 normal skin samples);
- growth condition (for example yeast cells cultured with different carbon sources)
- treatment (for example human cells treated with a given drug versus untreated control)
- Disease-vector insects (e.g. mosquito) that are either infected or non-infected by a parasite (e.g. the malaria agent, Plasmodium falciparum)
- Bacteria samples from very different environment
- ...
- ...


## Data sets


For each sample, biologists collected the RNA, obtained complementary DNA (cDNA), fragmented it, and sequencd the extremities of the fragments; The results is a file with short reads. 

The bioinformaicians then made a mapping of these short reads onto the reference genome, and compared the positions of the reads with the location of annotated genes. They counted the number of reads per gene in each sample. 


This is your input dataset:  a **count table** with one row per gene, one column per sample, and each value indicates the number of reads that were found for that gene in the sequencing results. 

## <font color="red">What is the main goal of differential expression analysis</font> ?
**Historical review**

<font color="red">The text below is merely a copy-oaste from a textbook. This should not be done without appropriate quotes and citation (if you don't quote the sentences, your document can be considered as plagiarism). </font>

"in the 1960s a consensus emerged  that cells differentiate through differential gene expression. The three postulates of differential gene expression are as follows:
1.Every cell nucleus contains the complete genome established in the fertilized egg. In molecular terms, the DNAs of all differentiated cells are identical.
2.The unused genes in differentiated cells are not destroyed or mutated, and they retain the potential for being expressed.
3.Only a small percentage of the genome is expressed in each cell, and a portion of the RNA synthesized in the cell is specific for that cell type." Source: Developmental Biology. 6th edition, free on-line edition <http://www.ncbi.nlm.nih.gov/books/NBK10061/>

You need now to show me that you understand the concepts, so please rephrase them with your **own** words. 

**Differential Expression ** 

"A gene is declared differentially expressed, if an observed difference or change in read counts between two experimental conditions is statistically significant, i.e. if the difference is greater than what would be expected just due to random
variation.
Statistical tools for microarrays were based on numerical intensity values.
Statistical tools for RNA-seq instead need to analyze read-count distributions." (please add the source OR suppress the citation)



*Goal** Detect genes whose transcripts (mRNAs) have changed significantly in abundance across different experimental conditions.  


**How ?** 

In general, we are looking into table that summarizes count data for each library (biological sample)  thence performing statistical testing between groups of samples of interest. Indeed, in order to evaluate if the difference between two experimental conditions is significant, it is not sufficient to dispose of a single biological sample per condition. Instead, we need to have as many **replicates** as possible for each condition. 


The biological concept of replicates corresponds to the statistical concept of "sample". Sorry for the confusion, we will sometimes use sample in the biological sense ( a single experiment, for example a piece of tissue, a drop of blood, ...), whereas for the statistician one sample includes the observations from different individuals. 

Example: psoriasis paper: 

the population is the set of all the people around the world that suffer from psoriasis. You can never obtain biological samples from these millions of people. The statistical sample is the subset of 92 persons suffering from psoriasis from whom Li and co-workers obtained a biological sample (i.e. a punch biopsis).















There are many methods have been used to develop the analysis of differentail expression like as microarray data and etc. But the RNA-Seq gives a discrete measurement for each gene, where trasformation of count data in not well approximated by countinuous distributions. espicailly in the lower count range and for small samples. therefore the appropriate statistical moleds for count dta are vital to extracting the most informaton from RNA-Seq Data.  


****************************************************************
# Fast reasoning

RNA-seq is based on the collection of RNA molecules from a biological sample.


We can consider the full set of RNA molecules found in the biological sample as the population (in the statistical sense) of all the transcripts produced by this sample at the time of the experiment. 

This RNA is then reverse transcribed to cDNA, because sequencing machines work with DNA, not RNA. The reverse transcription will already modify the relative representation of the different transcripts. At this stage, we could say this collection of cDNA is produced from a sampling of the mRNA (some mRNA molecules will be transcribed more efficiently than some other ones). 

There is also a DNA amplification step with PCR, which modifies the relative proportions of the different transcripts. 

The fragmentation is random, it cuts the full-length cDNA into pieces of approximately 300bp (typical size, but this can be adjusted by the experimenter). Generally, fragment lengths are distributed between ~200 and ~400bp. 

The fragments are then sequenced, but curently the sequencing is only partial: it is restricted to either one extremity (single-end) or both extremities (paired-ends).







