---
title: "Terminology and Concepts in statistical of RNA-Seq Data"
author: "Mustafa Abu El-Qumsan & Jacques van Helden"
date: '`r Sys.Date()`'
output:
  html_document:
    fig_caption: yes
    highlight: zenburn
    theme: cerulean
    toc: yes
    toc_depth: 3
    toc_float: true
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 3
  word_document: default
bibliography: Terminology.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Diffrential Expression With RNA-Seq Data

This method for DE Detection rely on a test statistics, used to identify which genes are characterized by a statistics significant change in gene expression in the compared conditions.
In principle, non-parametric methods can be used, But because of the small number of replicates typically available in RNA-Seq experiments, so that Non-parametric methods usually don't offer enough detection power, also parametric methods are preferred, owing to parametric method assumes a specific model to describe the underlying distribution od count data, furthermore seeks to identify the genes whose differences between the tested conditions exceed the variability pedicted by the model.
the models considered and implemented in the most of the analysis tools are based on the poisson and Negative Bionomial(NB) distributions.

<font color="blue"> in the past, we had presented a statistical description of the parameterization of RNA-Seq count data further that we need a more general summary of state-of-the-art approaches for DE analysis in RNA-Seq studies.</font>

## Models of RNA-Seq count data
The most used NB-based model of RNA-Seq counts is mean we have two parameters $\Phi$ and $\mu$
$$N_{f,j} \sim NB(\mu{f},\phi)$$
* where the parameter $\phi$ is \textbf{overdispersion}
The NB distribution can be motivated as Gamma mixture of poisson distribution- the technical variability is piosson, but the poisson means differ between biological replicates according to a Gamma distribution.

## Tools for DE analysis of RNA-Seq data
Given a specific statsitcal Model of RNA-Seq count data, all parametric tools for DE analysis cosist in two main steps:
1. Estimation of model parameters from data.
2. Detection of DE genes with a test statistics. 
where Library normalization can be considered part of DE analysis.

# Data exploration and quality assessment

## Data transformation

For data exploration and visualisation, it is useful to work with transformed versions of the count data. As the count values distribution is highly skewed, the *log 2 transformation* helps to approximately normalize the distributions.
Log base 2 is typically used as it facilitates the conversion back to the original scale: a difference of 1 on the log base 2 scale corresponds to a fold change of 2 on the original count scale. Since count values for a gene can be zero in some conditions (and non-zero in others), we advocate the use of raw counts, i.e. transformations of the form


$$y= \log_{2}(k+1) \text{   or more generally:   } y= \log_{2}(K + k_{0})$$

<font color="red">JvH: please find back the reference, and check if there is a difference between $k$ and  $K$. Also, always cite the source of all the information, this facilitates tractability. </font>

Even more, common statistical methods for exploratory analysis of multidimensional data, especially methods for clustering and ordination (e. g., principal-component analysis and the like), work best for (at least approximately) homoskedastic data; this means that the variance of an observable quantity (i.e., here, the expression strength of a
gene) does not depend on the mean. In RNA-Seq data, however, variance grows with the mean, with larger variances for larger counts. For example, if one performs PCA directly on a matrix of “normalized” read counts, the result typically depends only on the few most strongly expressed genes because they show the largest absolute differences between samples. A simple strategy to avoid this is to take the logarithm of the “normalized” count values plus a small constant; however, now the genes with low counts tend to dominate the results because, due to the strong “Poisson” noise inherent to small count values, they show the strongest relative differences between samples.


## box Plots

The boxplot method provides an easy way to visualize the distribution of read counts in each sample. In the
boxplot display, a box is formed with sides at the 25-th and 75-th percentiles of the distribution. A line is also drawn within the box at the level of the median. Whiskers are also drawn extending beyond each end of the box with points beyond the whiskers typically indicating outliers.

## Histograms and density plots

Read counts distributions can also be summarized by means of a histogram or density plot. Histograms provide
more detail by enabling, for example, the detection of a secondary mode in the distribution.

## MA-plot between samples

An MA-plot is a plot of log-fold change (M-values, i.e. the log of the ratio of level counts for each gene between two samples) against the log-average (A-values, i.e. the average level counts for each gene across the two samples).
The MA-plot is a useful to visualize reproducibility between samples of an experiment. From a MA-plot one can see if normalization is needed.
In MA plot, genes with similar expression levels in two samples will appear around the horizontal line y = 0. A low fit (in red) is plotted underlying a possible trend in the bias related to the mean expression.

## Clustering of the sample-to-sample distances

To explore the similarities and dissimilarities between samples, it is often instructive to look a clustering image map (CIM) or heatmap of sample-to-sample distance matrix.
A heatmap is a two-dimensional, rectangular, colored grid. It displays data that themselves come in the form of a
rectangular matrix:
– the color of each rectangle is determined by the value of the corresponding entry in the matrix,
– the rows and columns of the matrix are rearranged independently according to some hierarchical clustering
method, so that similar rows and columns are placed next to each other, respectively.

## Principal component plot of the samples

This type of plot is useful for visualizing the overall effect of experimental covariates and batch effects. The plot is produced from the output of Principal Componenet Analysis (PCA) on the transposed of the given counts matrix.
PCA is used to reduce multidimensional datasets to lower dimensions for analysis; it is a technique that can determine the key features of high-dimensional datasets. In the context of RNA-Seq analysis, PCA essentially clusters samples by groups of the most significantly dysregulated genes. Clustering first by the most significant group, then by progressively less significant groups.
Given the experimental design of the dataset that we are attempting to analyse here (i.e., samples belong to only two distinct groups), a wild-type group and a Mutant group, there should be a clear separation of both groups of samples by the first component.

## Multi-dimensional scaling plot
The purpose of multidimensional scaling (MDS) is to provide a visual representation of the pattern of proximities (i.e., similarities or distances) among a set of objects. MDS takes a set of similarities and returns a set of points such that the distances between the points are approximately equal to the similarities.
The similarities between each pair of samples (columns) is the root-mean-square deviation (Euclidean distance) for the top genes. Distances on the plot can be interpreted as leading log 2 -fold-change, meaning the typical (root-mean-square) log 2 -fold-change between the samples for the genes that distinguish those samples.
The function plotMDS has an argument gene.selection which, set to "common", chooses as top genes those with
the largest root-mean-square deviations between samples. Note D[i, j] the deviation between the samples i and j, and top how many of the most distinguished genes should be used in calculating the MDS,

# Raw data filtering

It is usually worthwhile to remove genes that appear to be very lowly expressed in any of the experimental conditions. This is called filtering.
There are a number of ways that filtering can be done. One way is to filter genes with a total read count smaller than a given threshold [@Sultan:2008:pmid18599741] and filter genes with at least one zero count in each experimental condition [@Bottomly:2011:pmid21455293].; however, selecting an arbitrary threshold value to filter genes in this way does not account for the overall sequencing depth.
To account differences in library size, filter has also been proposed based on counts per million (CPM) [@Robinson:2010:pmid19910308], calculated as the raw counts divided by the library sizes and multiplied by one million.
Genes with a CPM value less than a given cutoff (e.g. 1 or 100) in more samples (ignoring condition labels) than the size of the smallest group are subsequently filtered from the analysis.

# Normalization

Normalization is a process designed to identify and remove systematic technical differences between samples that occur in the data to ensure that technical bias has minimal impact on the results. The most common symptom of the need for normalization is differences in the total number of aligned reads.

## Total read count normalization

# References

